### DISCUSSION TITLE
Local development

### DISCUSSION THREAD

**Discussion by aviramha** [metalbear-co/mirrord#171] [General]
Hey everyone,
We've received few feedback from contributors about working locally.
I wanted to let you know we're aware of it and propose this as a discussion to suggest ideas on how to enable easier on boarding for new contributors.
It's quite funny that mirrord that should help testing and working on backend services lacks adequate ways to develop itself but that makes sense, as it's a quite complex architecture in terms of many components and parts creating the experience we want backend engineers to have.

This are the challenges we see in local development:

- **Running a K8S cluster** - We have minikube, but it isn't well supported for M1 and has some issues. if anyone has any idea for an alternative or maybe a "free" kubernetes service with good APIs that can be used for testing, that'd be great.
- **Running test agent image** - The agent comes as a docker image, and if we're running it remotely we need to push it to a remote registry, making it harder and more annoying to maintain. I wonder if there's a way to push an image to a node/k8s cluster so it can be used instead of a remote registry. Alternatively, if we make minikube/other work locally then this can be worked around with a script that builds + pushes to the local registry.
- **Running E2E** - This is the aggregation of the two above issues, the logic for setup/teardown can be part of the Rust test code or external, as a script or in the CI yaml. I guess the best solution would be either Rust/Bash so people can run it locally.

Would love to hear your thoughts and issues you saw as well.
Thanks!

**Comment by tamasfe** [comment]
I've used [kind](https://kind.sigs.k8s.io/) to test in-cluster services a lot. It supports loading images locally.

The cluster runs in a container so on Linux it gets its own IP, ingresses can be also done via metallb. On OSes where native docker is not supported (pretty much everywhere that is not Linux) networking setup could be non-trivial, but I have no experience with those.
